### ISSUE TITLE
Catch possibly undefined names in except clauses

### ISSUE THREAD

**Issue by afoglia** [astral-sh/ruff#23877]
### Summary

ruff version: 0.15.5

We stumbled across a case where the undefined name type checks can be improved.

Consider the case where we create an object, but also use it to handle an exception. For example,
```python
class Foo:
    def __init__(self):
        # init stuff that might throw

    def handle_exception(self):
        pass


try:
    foo = Foo()
    # other stuff that might throw
except Exception:
    foo.handle_exception()
```

In general, `Foo.__init__` won't always throw, but it can, in which case in the except-block, `foo` will be undefined. Ideally there would be a checker message on the last line.

`ruff check --preview --select ALL` does not report an issue. No checks currently seem to catch it.

Comparing to other checkers:
* pyflakes does not catch it
* pylint does ("E0601: Using variable 'foo' before assignment (used-before-assignment)")
* PyCharm's default linter does with the error "Local variable 'foo' might be referenced before assignment".

This seems like it should fall under F821, although because pyflakes doesn't catch that error, would perhaps copying over (PL)E0601 from pylint make more sense?

**Comment by ntBre** [comment]
I think this might be an intentional bias toward false negatives (over false positives) in Ruff. It looks like ty can detect this, but only with the off-by-default rule [possibly-unresolved-reference](https://docs.astral.sh/ty/reference/rules/#possibly-unresolved-reference). Even with ty's type inference, the rule is prone to false positives.

https://play.ty.dev/7ce4114f-eef4-48a3-b9a7-c59748c18596

**Comment by afoglia** [comment]
Can you give me an example of what you mean by a "false positive"? Because I can't think of one.

For it to be false, the initializer must be known to never throw, in which case it can be moved outside the try block with no ill effect. (That's what we did to silence it in PyCharm.)

If you wondering about a case like
```python
foo = Foo()

try:
  foo = Foo()
except Exception:
  foo.handle_exception()
```
both pylint and ty correctly don't warn about the block.

If you're concerned users may not care and want to silence it without silencing F802, perhaps a separate `possibily-undefined-rule` (or porting pylint's [E601 ("used before assignment")](https://pylint.readthedocs.io/en/latest/user_guide/messages/error/used-before-assignment.html) is the right approach.

PS: Writing this lead me to examine pylint's E601 definition, and it also catches stuff like this:
```python
try:
    foo = Foo()
except Exception:
    x = 10

print(x)
```
by reporting that `x` in the last statement is not necessarily assigned. That's an argument to just port E601 to ruff as PLE601, and leave F802 unchanged.

**Comment by MichaReiser** [comment]
Let's move this over to https://github.com/astral-sh/ruff/issues/16839, which asks for the same to have all conversations in a single place. @afoglia feel free to copy your comment over. I renamed https://github.com/astral-sh/ruff/issues/16839 to make it easier discoverable.
